modules/bag/src/zenoh_player.cpp

```diff
@@ -0,0 +1,191 @@
+  throwExceptionIf<InvalidOperationException>(terminate_, "player is already stopped, cannot stop again");
+  terminate_ = true;
+  play_cv_.notify_all();
+
+  return std::async(std::launch::async, [this]() { run_job_.get(); });
+}
+
+void ZenohPlayer::Impl::wait() const {
+  run_job_.wait();
+}
+
+void ZenohPlayer::Impl::createPublisher(const mcap::Channel& channel) {
+  throwExceptionIf<InvalidDataException>(
+      publishers_.contains(channel.topic),
+      fmt::format("failed to create publisher for topic: {}; topic already exist", channel.topic));
+
+  const auto& schema = bag_reader_->schema(channel.schemaId);
+  auto type_info = serdes::TypeInfo{
+    .name = schema->name,
+    .schema = schema->data,
+    .serialization = serdes::TypeInfo::Serialization::PROTOBUF,  // TODO: get this from schema
+    .original_type =
+        "",  // TODO: figure out if there is a way to get this. We can use the channel metadata!!!
+  };
+
+  publishers_[channel.topic] = std::make_unique<ipc::zenoh::Publisher>(
+      session_, ipc::TopicConfig{ .name = channel.topic }, std::move(type_info),
+      [this, &channel](ipc::zenoh::MatchingStatus status) {
+        if (!status.matching) {
+          return;
+        }
+
+        publishers_with_subscriber_.insert(channel.topic);
+        if (publishers_with_subscriber_.size() == channel_count_) {
+          all_publisher_connected_.test_and_set();
+          all_publisher_connected_.notify_all();
+        }
+      });
+
+  LOG(INFO) << fmt::format("Created publisher for topic: {}", channel.topic);
+}
+
+void ZenohPlayer::Impl::run() {
+  auto read_options = mcap::ReadMessageOptions{};
+  read_options.readOrder = mcap::ReadMessageOptions::ReadOrder::LogTimeOrder;
+  auto messages = bag_reader_->readMessages([](const auto&) {}, read_options);
+
+  auto first_msg_timestamp = std::chrono::steady_clock::time_point{ std::chrono::nanoseconds{
+      messages.begin()->message.publishTime } };
+
+  std::size_t msgs_played_count = 0;
+  std::size_t deadline_missed_count = 0;
+
+  LOG_IF(WARNING, wait_for_readers_to_connect_) << "Waiting for subscribers to connect is NOT supported yet!";
+  if (wait_for_readers_to_connect_) {
+    all_publisher_connected_.wait(false);
+  }
+
+  const auto first_playback_timestamp = std::chrono::steady_clock::now();
+  for (const auto& message : messages) {
+    if (terminate_) {
+      break;
+    }
+
+    const auto& topic = message.channel->topic;
+    const auto& publisher = publishers_[topic];
+
+    const auto current_msg_timestamp =
+        std::chrono::steady_clock::time_point{ std::chrono::nanoseconds{ message.message.publishTime } };
+
+    const auto write_timestamp = current_msg_timestamp - first_msg_timestamp + first_playback_timestamp;
+
+    if (auto now = std::chrono::steady_clock::now(); now > write_timestamp && msgs_played_count > 0) {
+      ++deadline_missed_count;
+      LOG(WARNING) << fmt::format("failed to publish message {} in time for topic {}, delay {}",
```

Comment:
I would change this error message, to `deadline missed on message...`

Reply:
nice!